internal/pkg/uploader/finalize.go

```diff
@@ -0,0 +1,151 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package uploader
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"sort"
+	"strings"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/uploader/upload"
+	"github.com/rs/zerolog/log"
+)
+
+var (
+	ErrFailValidation  = errors.New("file contents failed validation")
+	ErrStatusNoUploads = errors.New("file closed, not accepting uploads")
+)
+
+func (u *Uploader) Complete(ctx context.Context, id string, transitHash string) (upload.Info, error) {
+	// make sure document is freshly fetched, not cached
+	// so accurate status checking happens
+	info, err := FetchUploadInfo(ctx, u.bulker, id)
+	if err != nil {
+		return info, err
+	}
+
+	/*
+		Verify Upload
+	*/
+
+	// if already done, failed or deleted, exit
+	if !info.StatusCanUpload() {
+		return info, ErrStatusNoUploads
+	}
+
+	chunks, err := GetChunkInfos(ctx, u.bulker, info.DocID)
+	if err != nil {
+		return info, err
+	}
+	if !u.allChunksPresent(info, chunks) {
+		return info, ErrMissingChunks
+	}
+	if !u.verifyChunkInfo(info, chunks, transitHash) {
+		if err := SetStatus(ctx, u.bulker, info, upload.StatusFail); err != nil {
+			log.Warn().Err(err).Str("fileID", info.DocID).Str("uploadID", info.ID).Msg("file upload failed chunk validation, but encountered an error setting the upload status to failure")
+		}
+		if err := DeleteChunksByQuery(ctx, u.bulker, info.Source, info.DocID); err != nil {
+			log.Warn().Err(err).Str("fileID", info.DocID).Str("uploadID", info.ID).Msg("file upload failed chunk validation, but encountered an error deleting left-behind chunk data")
+		}
+		return info, ErrFailValidation
+	}
+
+	/*
+		Upload OK. Update status and save valid transithash
+	*/
+	if err := MarkComplete(ctx, u.bulker, info, transitHash); err != nil {
+		return info, err
+
+	}
+
+	return info, nil
+}
+
+func (u *Uploader) allChunksPresent(info upload.Info, chunks []ChunkInfo) bool {
+	// check overall count
+	if len(chunks) != info.Count {
+		log.Warn().Int("expectedCount", info.Count).Int("received", len(chunks)).Interface("chunks", chunks).Msg("mismatch number of chunks")
+		return false
+	}
+
+	// now ensure all positions are accounted for, no gaps, etc
+	sort.Slice(chunks, func(i, j int) bool {
```

Comment:
question/nit: could you avoid sorting if the chunks are always selected "ordered by pos" from elasticsearch?

Reply:
I'm not sure I can guarantee that they are always sorted correctly by ES. The positional information is not in a sortable field, it is in just a portion of the document's `_id`. Considering this is a bounded size slice (limited maximum file size), sorting was not out of the question